### PR TITLE
Ensure that angle.to_string() continues to work after pickling

### LINKED ISSUE #13923: Unpickled Angle.to_string fails
### Description

Pickling and unpickling an Angle object causes the to_string function to fail claiming hourangle and degree units cannot be represented in sexagesimal notation.

### Steps to Reproduce

```python
import astropy.coordinates
import pickle
ang = astropy.coordinates.Angle(0.25 * astropy.units.hourangle)
pang = pickle.loads(pickle.dumps(ang))
ang.to_string()
# Works: 0h15m00s
pang.to_string()
# Fails: ValueError: 'hourangle' can not be represented in sexagesimal notation
```

### System Details

macOS-10.15.7-x86_64-i386-64bit
Python 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:43:44) [Clang 13.0.1 ]
Numpy 1.23.4
pyerfa 2.0.0.1
astropy 5.1
Scipy 1.9.3
Matplotlib 3.6.1

### PR DESCRIPTION
**Pull request by mhvk**
### Description

This pull request is to address that after pickling `angle.to_string()` did not work any more. In the process, it also ensures that composite units like `u.deg ** 1` are valid as well.

Fixes #13923

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label. Codestyle issues can be fixed by the [bot](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit).
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by mhvk** [comment]
Oops, failures are real (in `wcsaxis`, pinging @astrofrog as this seems weird...)

**Comment by astrofrog** [comment]
We do use Angle formatting quite a lot in WCSAxes so the failures are likely indeed real with this PR. I won't have time to investigate this for a few days but if you want to get this into 5.2 feel free to try and investigate - this is happening in the formatter and locator code which isn't too complex and doesn't actually do any Matplotlib stuff.

**Comment by pllim** [comment]
Looks to be a matter of "d" vs the degree symbol. Is it some sort of LaTeX formatting not happening with this change?

```
FAILED .../visualization/wcsaxes/tests/test_display_world_coordinates.py::
TestDisplayWorldCoordinate::test_overlay_coords -
assert "267d -28°45'...d, overlay 1)" == '267.176° -28...d, overlay 1)'
  - 267.176° -28°45'56" (world, overlay 1)
  ?    ^^^^^
  + 267d -28°45'56" (world, overlay 1)
  ?    ^
FAILED .../visualization/wcsaxes/tests/test_formatter_locator.py::
TestAngleFormatterLocator::test_format[d-15\xb0] - 
AssertionError: assert '15d' == '15°'
  - 15°
  + 15d
```

**Comment by mhvk** [comment]
The problem is that `wcsaxis` makes use of the identity check: https://github.com/astropy/astropy/blob/0c37a7141c6c21c52ce054f2d895f6f6eacbf24b/astropy/visualization/wcsaxes/formatter_locator.py#L396-L402

**Comment by mhvk** [comment]
Fixing this was a bit more involved than I hoped. To me, it is very strange that by default `to_string(decimal=True)` does not include a unit at all, which is what is causing the problems in `wcsaxes`. I slightly cheated by now making it such that a unit will be added if any `format` is passed on (previously, `format` was ignored for this case). I'm not sure this is quite OK; frankly, I wish `decimal=True` by default would just use `Quantity.to_string()`... Anyway, what do people think? Can this go in? Should it still be backported?

**Review by astrofrog — APPROVED** [review]
I'm happy with the WCSAxes part of the fix but have not reviewed in detail the changes to astropy.coordinates

I think this can still be included in 5.2 and backported to v5.0.x - if possible we should finalize/merge this by end of Tuesday to prevent conflicts with the Black PRs.

**Comment by mhvk** [comment]
OK, that sounds good. I added a few more tests for the coordinates changes.

**Comment by astrofrog** [comment]
@mhvk - do you want any reviews from other maintainers or do you think we can go ahead and merge?

**Comment by mhvk** [comment]
@astrofrog - I think we can get it in; it is certainly fine for 5.2. Not 100% OK for the backport since it adds somewhat of a feature as part of the solution, but really I do not see that becoming a problem.

Note that I cannot merge since this is part feature freeze.

**Pull request merged**
